block-producer/src/txqueue/mod.rs

```diff
@@ -0,0 +1,125 @@
+use std::{fmt::Debug, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use tokio::{sync::RwLock, time};
+
+use crate::{batch_builder::BatchBuilder, SharedProvenTx, SharedRwVec};
+
+// TRANSACTION QUEUE
+// ================================================================================================
+
+#[async_trait]
+pub trait TransactionVerifier: Send + Sync + 'static {
+    type VerifyTxError: Debug;
+
+    async fn verify_tx(
+        &self,
+        tx: SharedProvenTx,
+    ) -> Result<(), Self::VerifyTxError>;
+}
+
+#[async_trait]
+pub trait TransactionQueue: Send + Sync + 'static {
+    type AddTransactionError;
+
+    async fn add_transaction(
+        &self,
+        tx: SharedProvenTx,
+    ) -> Result<(), Self::AddTransactionError>;
+}
+
+pub enum AddTransactionError {
+    VerificationFailed,
+}
+
+// DEFAULT TRANSACTION QUEUE
+// ================================================================================================
+
+pub struct DefaultTransactionQueueOptions {
+    /// The frequency at which we try to send transaction groups
+    pub send_tx_groups_frequency: Duration,
+
+    /// The size of a batch
+    pub batch_size: usize,
+}
+
+pub struct DefaultTransactionQueue<BB: BatchBuilder, TV: TransactionVerifier> {
+    ready_queue: SharedRwVec<SharedProvenTx>,
+    tx_verifier: Arc<TV>,
+    batch_builder: Arc<BB>,
+    options: DefaultTransactionQueueOptions,
+}
+
+impl<BB, TV> DefaultTransactionQueue<BB, TV>
+where
+    TV: TransactionVerifier,
+    BB: BatchBuilder,
+{
+    pub fn new(
+        tx_verifier: Arc<TV>,
+        batch_builder: Arc<BB>,
+        options: DefaultTransactionQueueOptions,
+    ) -> Self {
+        Self {
+            ready_queue: Arc::new(RwLock::new(Vec::new())),
+            tx_verifier,
+            batch_builder,
+            options,
+        }
+    }
+
+    pub async fn run(self) {
+        let mut interval = time::interval(self.options.send_tx_groups_frequency);
+
+        loop {
+            interval.tick().await;
+            self.try_send_tx_groups().await;
+        }
+    }
+
+    async fn try_send_tx_groups(&self) {
+        let mut locked_ready_queue = self.ready_queue.write().await;
+
+        if locked_ready_queue.is_empty() {
+            return;
+        }
+
+        let tx_groups: Vec<Vec<SharedProvenTx>> = locked_ready_queue
+            .chunks(self.options.batch_size)
+            .map(|txs| txs.to_vec())
+            .collect();
+
+        match self.batch_builder.add_tx_groups(tx_groups).await {
+            Ok(_) => {
+                // Transaction groups were successfully sent, so drain the queue
+                locked_ready_queue.truncate(0);
+            },
+            Err(_) => {
+                // Transaction groups were not sent, and remain in the queue. Do nothing.
+            },
+        }
```

Comment:
As mentioned in one of the previous comments, I think it might be better to send batches individually rather than as a whole group. This should give us better control over failures.

Also, I think there might be an issue here: we are removing transactions from the queue only after batches have been successfully built. But what if it takes a while to build a batch (e.g., 2 seconds)? Wouldn't `try_send_tx_groups()` be called again on the timer and then we'd send the same set of transactions for batching again?

Maybe the following would be a better approach?:
- We remove transactions to be batched from the queue right away.
- We send the transaction groups to the batch builder and await the results for each group separately.
- If a group was successfully put into a batch, we do nothing.
- But if there was a failure we re-insert the transactions into a batch.
  - For this, we'll need to know which transaction group failed. One way to do this is to return the list of transactions with the error. But there could be a better way to do it as well.

One open question in my mind (which we don't need to solve in this PR, but would be good to think through potential solutions) is how to handle timeouts. Specifically, if a batch is taking too long to build (e.g., takes 5 seconds), we should abort and re-insert its transactions into the queue.

Reply:
> Wouldn't try_send_tx_groups() be called again on the timer and then we'd send the same set of transactions for batching again?

No because we are [holding the lock](https://github.com/0xPolygonMiden/miden-node/blob/60691e44717e0407d13c4607ad399a117e399689/block-producer/src/txqueue/mod.rs#L81) on the `ready_queue` the entire time.

I did it this way so that transactions that arrived earlier make it into a block before transactions that arrive later (but this is another one of my random guesses at whether or not this is a requirement). So
1. `try_send_tx_groups()` cannot be called if another call is still in progress because [we await it](https://github.com/0xPolygonMiden/miden-node/blob/60691e44717e0407d13c4607ad399a117e399689/block-producer/src/txqueue/mod.rs#L76). 
2.But even if we could, we're holding the lock for the entire duration of `try_send_tx_groups()`, so the second call would just block there


I agree though that if a single transaction group fails, the other ones should be able to proceed and be put into a block.